lib/Core/Site/QueryType/CriterionDefinitionResolver.php

```diff
@@ -0,0 +1,208 @@
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @param string $name
+     * @param string|null $target
+     * @param mixed $parameters
+     *
+     * @return \Netgen\EzPlatformSiteApi\Core\Site\QueryType\CriterionDefinition[]
+     */
+    private function resolveForTarget($name, $target, $parameters)
+    {
+        if ($this->isOperatorMap($parameters)) {
+            return $this->resolveOperatorMap($name, $target, $parameters);
+        }
+
+        return [
+            $this->buildDefinition($name, $target, null, $parameters),
+        ];
+    }
+
+    /**
+     * Return CriterionDefinition instances for the given Field $target and its operator $map.
+     *
+     * @param string $name
+     * @param string|null $target
+     * @param array $map
+     *
+     * @return \Netgen\EzPlatformSiteApi\Core\Site\QueryType\CriterionDefinition[]
+     */
+    private function resolveOperatorMap($name, $target, array $map)
+    {
+        $definitions = [];
+
+        foreach ($map as $operator => $value) {
+            if ('not' === $operator) {
+                $definitions[] = $this->buildDefinition(
+                    'not',
+                    null,
+                    null,
+                    $this->resolveForTarget($name, $target, $value)
+                );
+            } else {
+                $definitions[] = $this->buildDefinition($name, $target, $operator, $value);
+            }
+        }
+
+        return $definitions;
+    }
+
+    /**
+     * Return CriterionDefinition instance from the given arguments.
+     *
+     * @param string $name
+     * @param string|null $target
+     * @param string|null $operator
+     * @param mixed $value
+     *
+     * @return \Netgen\EzPlatformSiteApi\Core\Site\QueryType\CriterionDefinition
+     */
+    private function buildDefinition($name, $target, $operator, $value)
+    {
+        return new CriterionDefinition([
+            'name' => $name,
+            'target' => $target,
+            'operator' => $this->resolveOperator($operator, $value),
+            'value' => $value,
+        ]);
+    }
+
+    /**
+     * Decide if the given $parameters is an operator-value map.
+     *
+     * @param mixed $parameters
+     *
+     * @throws \InvalidArgumentException
+     *
+     * @return bool
+     */
+    private function isOperatorMap($parameters)
+    {
+        if (!is_array($parameters)) {
+            return false;
+        }
+
+        $isOperatorMap = false;
+        $isValueCollection = false;
+
+        foreach (array_keys($parameters) as $key) {
+            if (array_key_exists($key, self::$operatorMap) || 'not' === $key) {
+                $isOperatorMap = true;
+            } else {
+                $isValueCollection = true;
+            }
+        }
+
+        if ($isOperatorMap && $isValueCollection) {
+            throw new InvalidArgumentException(
+                'Array of parameters is not an operator map nor a value collection'
```

Comment:
Maybe the exception message is a little bit misleading? :)

Reply:
It is :)

Improved a little bit in https://github.com/netgen/ezplatform-site-api/pull/70/commits/ce90b97c475b5d008d4481a88dfe228184509b70